### ISSUE TITLE
Unable to fetch all balances for a sif account with large number of denoms

### ISSUE THREAD

**Issue by jzvikart** [Sifchain/sifnode#2858]
To reproduce:
1. Create a sif wallet with 100000 denoms. For some reason, it is important that you create it with tokens coming from Ethereum, not from genesis.
2. `sifnoded query bank balances --offset 99999 --limit 1`

Result:
```
Error: post failed: Post "http://localhost:26657": EOF
```

Expected result: something like
```
balances:
- amount: "123456"
  denom: sifBridge99990x0002273068a986d63f0367f4cab49c1aaf1e3595
pagination:
  next_key: c2lmQnJpZGdlOTk5OTB4MDAwMjMzZDg2NDA2ZTJmYjg3M2MxZjNjYzllODFhZWVlYzE1NDFlOQ==
  total: "0"
```

Differential diagnosis: `sifnoded query bank balances --limit 1 --offset 1` works OK.

Related issues:
- #2520
- #2794

**Comment by jzvikart** [comment]
The time needed to access n-th element of balance seems to be O(n).

- `query bank balances --limit 1 --offset 0`: 0.07s
- `query bank balances --limit 1 --offset 20000`: 3.3s
- `query bank balances --limit 1 --offset 40000`: 6,9s
- `query bank balances --limit 1 --offset 60000`: 10.2s
- `query bank balances --limit 1 --offset 80000`: `Error: post failed: Post "http://localhost:26657": EOF`

This means that paging not only does not solve the problem, but in a way makes things worse:
- without paging you would need 1 request which is O(n)
- with paging you need O(n) requests where of them takes O(N) time => O(N^2) total time

**Comment by pandaring2you** [comment]
This is a cosmos SDK issue. Objective now is to figure out the limits on number of token balances and/or figure out if there are workarounds.

**Comment by jzvikart** [comment]
As suggested by @timlind, when specifying offset with `--page-key` instead of `--offset`, there is no seek penalty and the total time is O(N).

However, timeouts are still happening, and they happen in a non-deterministic manner. In other words, the same request sometimes works and sometimes fails. My impression is that the sucess depends on the speed with which the server can fulfill the request.

**Comment by jzvikart** [comment]
@timlind I would like to try to try with a longer timeout, how can I change it?

**Comment by timlind** [comment]
What do you mean by timeout?

**Comment by jzvikart** [comment]
`sifnoded query bank balances` is obviously making a http request in the backgroud. This request times out with large balances, so I would like to try to increase the timeout to see if that solves the problem.

**Comment by timlind** [comment]
I don't know of anyway to control that.

**Comment by pandaring2you** [comment]
Brando: This is a cosmos sdk issue. We should raise an issue with the cosmos team and determine if we should spend our own resources to solve this. 

@jzvikart - try to run test with denom flag and size = 1 to ensure cosmos does not have deeper issues.  

@smartyalgo @Brando753 - Can you file the issue Brandon mentioned with Cosmos when you identify the part of the bank module code that is problematic and link it to this task?  

jure sidenote: noticed timeouts usually happen when a new block is being mined

**Comment by jzvikart** [comment]
I cannot test this properly until the bridge is working as it should (not hanging + not filling up the disks with logs + decent performance).

**Comment by pandaring2you** [comment]
@Brando753 @timlind - Did you have a hunch as to where this issue lies in the bank module code? Based on my Jun 7th comment above, it seemed like you or @smartyalgo might have a hunch as to what is causing this issue. If not, perhaps we can close this after the TPS fix is in and tested with Jon's approval as this poses an unknown production risk.

**Comment by Brando753** [comment]
I have not dug into the bank module code at all. The hunch was that this was probably related to how the bank module stores users funds in the keeper. Agreed probably makes sense to look into after the TPS fix.

**Comment by pandaring2you** [comment]
Note: Need a more scalable solution for tps < 1, but unblocked on this test.

**Comment by jzvikart** [comment]
@pandaring2you In the last round of tests I did with 100k wallets I was getting a lot of errors and timeouts. I added retry logic to my code which eventually made it possible for tests to read all the balances (albeit taking very long time). As far as this ticket is concerned we can close it, but the open questions about performance and scalability still remain, and they are what really matters. To answer those, we would need to do more load testing,
